src/equations/shallow_water_multilayer_2d.jl

```diff
@@ -0,0 +1,827 @@
+    momentum_1 = ntuple(n -> "h" * string(n) * "_v1", Val(nlayers(equations)))
+    momentum_2 = ntuple(n -> "h" * string(n) * "_v2", Val(nlayers(equations)))
+    return (waterheight..., momentum_1..., momentum_2..., "b")
+end
+
+# We use the total layer heights, H = ∑h + b as primitive variables for easier visualization and setting initial
+# conditions
+function Trixi.varnames(::typeof(cons2prim),
+                        equations::ShallowWaterMultiLayerEquations2D)
+    total_layer_height = ntuple(n -> "H" * string(n), Val(nlayers(equations)))
+    velocity_1 = ntuple(n -> "v" * string(n) * "_1", Val(nlayers(equations)))
+    velocity_2 = ntuple(n -> "v" * string(n) * "_2", Val(nlayers(equations)))
+    return (total_layer_height..., velocity_1..., velocity_2..., "b")
+end
+
+# # Set initial conditions at physical location `x` for time `t`
+"""
+    initial_condition_convergence_test(x, t, equations::ShallowWaterMultiLayerEquations2D)
+
+A smooth initial condition for a three-layer configuration used for convergence tests in combination with
+[`source_terms_convergence_test`](@ref) (and
+[`Trixi.BoundaryConditionDirichlet`](@extref) in non-periodic domains).
+"""
+function Trixi.initial_condition_convergence_test(x, t,
+                                                  equations::ShallowWaterMultiLayerEquations2D)
+    # Check if the equation has been set up with three layers
+    if nlayers(equations) != 3
+        throw(ArgumentError("This initial condition is only valid for three layers"))
+    end
+
+    # Some constants are chosen such that the function is periodic on the domain [0,sqrt(2)]
+    ω = 2.0 * pi * sqrt(2.0)
+
+    H3 = 1.5 + 0.1 * cos(ω * x[1] + t) + 0.1 * cos(ω * x[2] + t)
+    H2 = 2.0 + 0.1 * sin(ω * x[1] + t) + 0.1 * sin(ω * x[2] + t)
+    H1 = 4.0 + 0.1 * cos(ω * x[1] + t) + 0.1 * cos(ω * x[2] + t)
+    H = (H1, H2, H3)
+    v1 = (0.8, 0.8, 0.8)
+    v2 = (1.0, 1.0, 1.0)
+
+    b = 1.0 + 0.1 * cos(ω * x[1]) + 0.1 * cos(ω * x[2])
+
+    return prim2cons(SVector(H..., v1..., v2..., b), equations)
+end
+
+"""
+    source_terms_convergence_test(u, x, t, equations::ShallowWaterMultiLayerEquations2D)
+
+Source terms used for convergence tests with a three-layer configuration in combination with
+[`initial_condition_convergence_test`](@ref)
+(and [`Trixi.BoundaryConditionDirichlet`](@extref)
+in non-periodic domains).
+"""
+@inline function Trixi.source_terms_convergence_test(u, x, t,
+                                                     equations::ShallowWaterMultiLayerEquations2D)
+    # Same settings as in `initial_condition_convergence_test`. Some derivative simplify because
+    # this manufactured solution velocity is taken to be constant
+    ω = 2 * pi * sqrt(2.0)
+
+    du1 = -0.1cos(t + x[2] * ω) - 0.1sin(t + x[1] * ω) - 0.1sin(t + x[2] * ω) -
+          0.1cos(t + x[1] * ω) - 0.1cos(t + x[2] * ω) * ω +
+          0.8(-0.1sin(t + x[1] * ω) * ω - 0.1cos(t + x[1] * ω) * ω) -
+          0.1sin(t + x[2] * ω) * ω
```

Comment:
Are some of these multiplications implicitly done, e.g., `0.1sin(t + x[2] * ω)`?

Reply:
Yes, here is the respective section in the [julia docs](https://docs.julialang.org/en/v1/manual/integers-and-floating-point-numbers/#man-numeric-literal-coefficients).
We don't typically use this in the code, but I think it actually improves the readability as the additional ` * ` would bloat the expression even more.